tests/unit/utils/SGObjectIterator.h

```diff
@@ -0,0 +1,164 @@
+#ifndef __SG_OBJECT_ITERATOR_H__
+#define __SG_OBJECT_ITERATOR_H__
+
+#include <iterator>
+#include <set>
+#include <string>
+#include <shogun/base/class_list.h>
+
+using namespace shogun;
+
+// to have a type for non-template SGObject classes
+struct untemplated_sgobject
+{
+};
+
+/** Returns primitive type value for template parameter */
+template <class T>
+static EPrimitiveType sg_primitive_type();
+template <class T>
+static std::string sg_primitive_type_string();
+
+// stringizing the result of expansion of a macro argument used below
+// https://gcc.gnu.org/onlinedocs/cpp/Stringizing.html
+#define STRING__(s) #s
+#define XSTRING__(s) STRING__(s)
+
+#define SG_PRIMITIVE_TYPE(T, ptype)                                            \
+	template <>                                                                \
+	EPrimitiveType sg_primitive_type<T>()                                      \
+	{                                                                          \
+		return ptype;                                                          \
+	}                                                                          \
+                                                                               \
+	template <>                                                                \
+	std::string sg_primitive_type_string<T>()                                  \
+	{                                                                          \
+		return std::string(XSTRING__(ptype));                                  \
+	}
+
+SG_PRIMITIVE_TYPE(bool, PT_BOOL);
```

Comment:
@gf712 do we have a central map for these?

Reply:
all the supported types should go here
https://github.com/shogun-toolbox/shogun/blob/801bdbaee76c95ff77a243a8717f0ecefbf66dc0/src/shogun/lib/sg_types.h#L41
and all the any supported types are here
https://github.com/shogun-toolbox/shogun/blob/801bdbaee76c95ff77a243a8717f0ecefbf66dc0/src/shogun/lib/type_case.h#L118

Reply:
@gf712 `DataType.h` seems to be obsolete, but I can't use `sg_types` and `type_case` yet since the `class_list` script (which has the `create` function) uses the old code.